src/algebra/group_with_zero.lean

```diff
@@ -0,0 +1,504 @@
+class comm_group_with_zero (G : Type*) extends comm_monoid G, group_with_zero G.
+
+/-- The division operation on a group with zero element. -/
+instance group_with_zero.has_div {G : Type*} [group_with_zero G] :
+  has_div G := ⟨λ g h, g * h⁻¹⟩
+
+end prio
+
+lemma div_eq_mul_inv {G : Type*} [group_with_zero G] {a b : G} :
+  a / b = a * b⁻¹ := rfl
+
+namespace gwz
+variables {G : Type*} [group_with_zero G]
+
+@[simp] lemma inv_zero : (0 : G)⁻¹ = 0 :=
+group_with_zero.inv_zero G
+
+@[simp] lemma mul_inv_cancel (a : G) (h : a ≠ 0) : a * a⁻¹ = 1 :=
+group_with_zero.mul_inv_cancel a h
+
+@[simp] lemma mul_inv_cancel_assoc_left (a b : G) (h : b ≠ 0) :
+  (a * b) * b⁻¹ = a :=
+calc (a * b) * b⁻¹ = a * (b * b⁻¹) : mul_assoc _ _ _
+               ... = a             : by simp [h]
+
+@[simp] lemma mul_inv_cancel_assoc_right (a b : G) (h : a ≠ 0) :
+  a * (a⁻¹ * b) = b :=
+calc a * (a⁻¹ * b) = (a * a⁻¹) * b : (mul_assoc _ _ _).symm
+               ... = b             : by simp [h]
+
+lemma inv_ne_zero {a : G} (h : a ≠ 0) : a⁻¹ ≠ 0 :=
+assume a_eq_0, by simpa [a_eq_0] using mul_inv_cancel a h
+
+@[simp] lemma inv_mul_cancel (a : G) (h : a ≠ 0) : a⁻¹ * a = 1 :=
+calc a⁻¹ * a = (a⁻¹ * a) * a⁻¹ * a⁻¹⁻¹ : by simp [inv_ne_zero h]
+         ... = a⁻¹ * a⁻¹⁻¹             : by simp [h]
+         ... = 1                       : by simp [inv_ne_zero h]
+
+@[simp] lemma inv_mul_cancel_assoc_left (a b : G) (h : b ≠ 0) :
+  (a * b⁻¹) * b = a :=
+calc (a * b⁻¹) * b = a * (b⁻¹ * b) : mul_assoc _ _ _
+               ... = a             : by simp [h]
+
+@[simp] lemma inv_mul_cancel_assoc_right (a b : G) (h : a ≠ 0) :
+  a⁻¹ * (a * b) = b :=
+calc a⁻¹ * (a * b) = (a⁻¹ * a) * b : (mul_assoc _ _ _).symm
+               ... = b             : by simp [h]
+
+@[simp] lemma inv_inv (a : G) : a⁻¹⁻¹ = a :=
+begin
+  classical,
+  by_cases h : a = 0, { simp [h] },
+  calc a⁻¹⁻¹ = a * (a⁻¹ * a⁻¹⁻¹) : by simp [h]
+        ... = a                 : by simp [inv_ne_zero h]
+end
+
+lemma inv_involutive : function.involutive (has_inv.inv : G → G) :=
+inv_inv
+
+lemma ne_zero_of_mul_right_eq_one (a b : G) (h : a * b = 1) : a ≠ 0 :=
+assume a_eq_0, zero_ne_one (by simpa [a_eq_0] using h : (0:G) = 1)
+
+lemma ne_zero_of_mul_left_eq_one (a b : G) (h : a * b = 1) : b ≠ 0 :=
+assume b_eq_0, zero_ne_one (by simpa [b_eq_0] using h : (0:G) = 1)
+
+lemma eq_inv_of_mul_right_eq_one (a b : G) (h : a * b = 1) :
+  b = a⁻¹ :=
+calc b = a⁻¹ * (a * b) :
+        (inv_mul_cancel_assoc_right _ _ $ ne_zero_of_mul_right_eq_one a b h).symm
+   ... = a⁻¹ : by simp [h]
+
+lemma eq_inv_of_mul_left_eq_one (a b : G) (h : a * b = 1) :
+  a = b⁻¹ :=
+calc a = (a * b) * b⁻¹ : (mul_inv_cancel_assoc_left _ _ (ne_zero_of_mul_left_eq_one a b h)).symm
+   ... = b⁻¹ : by simp [h]
+
+@[simp] lemma inv_one : (1 : G)⁻¹ = 1 :=
+eq.symm $ eq_inv_of_mul_right_eq_one _ _ (mul_one 1)
+
+lemma inv_injective : function.injective (@has_inv.inv G _) :=
```

Comment:
This follows from `injective_of_equiv` or whatever, if you define the equiv. Maybe it should also be a one liner from `inv_involutive` ?

Reply:
Fixed. Thanks.